### PR TITLE
Curl => 7.78.0

### PR DESCRIPTION
**Pull request by saltedcoffii**
- [x] x86_64
- [ ] i686
- [ ] armv7l/aarch64

Also upgraded supporting libraries.

### CODE REVIEW

**Comment by satmandu** [comment]
Nice. Setting up builds for x86_64 & armv7l.

**Comment by satmandu** [comment]
I think we might want to use the source links and not git links, because otherwise we see this:
```
chronos@cheekon-x86_64 /usr/local/lib/crew/packages $ curl --version
curl 7.78.0-DEV (x86_64-pc-linux-musl) libcurl/7.78.0-DEV OpenSSL/1.1.1k zlib/1.2.11 brotli/1.0.9 zstd/1.5.0 libidn2/2.3.2 nghttp2/1.44.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https mqtt pop3 pop3s smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets zstd
```

**Comment by satmandu** [comment]
Also metalink linking appears to be deprecated:
`configure: error: --with-libmetalink no longer works!`

File: packages/curl.rb [hunk 1]
```diff
@@ -3,40 +3,69 @@
...
+    Dir.chdir 'build/libssh' do
+      system 'curl -#L https://www.libssh.org/files/0.9/libssh-0.9.5.tar.xz | \
+       hashpipe sha256 acffef2da98e761fc1fd9c4fddde0f3af60ab44c4f5af05cd1b2d60a3fa08718 | \
+        tar xJ --strip-components=1'
+    end
+    Dir.chdir 'build/nghttp2' do
+      system 'curl -#L https://github.com/nghttp2/nghttp2/releases/download/v1.44.0/nghttp2-1.44.0.tar.bz2 | \
+        hashpipe sha256 989971276517a1c9ed330b779c34cf02d99da3b85d156eb297f42b1b7227b297 | \
+        tar xj --strip-components=1'
+    end
```
Comment (uberhacker) [inline packages/curl.rb:63]: Why not just use crew packages instead?

Comment (satmandu) [reply packages/curl.rb:63]: Because these are for building against musl...

**Comment by satmandu** [comment]
Added some binaries which need testing as I have no native hardware at the moment.

@uberhacker i686 binaries please?

**Comment by satmandu** [comment]
Whoops! Didn't include an armv7l build for libcurl. Starting that now.

**Comment by satmandu** [comment]
ok, all armv7l binaries are now uploaded (but unchecked) for this.

**Comment by uberhacker** [comment]
Overnight i686 build locked up at:
```
checking if getifaddrs can be linked... yes
checking if getifaddrs is prototyped... yes
checking if getifaddrs is compilable... yes
checking if getifaddrs seems to work... 
```

**Comment by uberhacker** [comment]
On armv7l, I get:
```
$ crew search curl -v
curl: Command line tool and library for transferring data with URLs.
https://curl.se/
Version: 7.78.0
License: curl
$ curl -L# git.io/vddgY -o install.sh
bash: /usr/local/bin/curl: No such file or directory
$ ls -l /usr/local/bin/curl
-rwxr-xr-x 1 chronos chronos 3514344 Aug 10 01:53 /usr/local/bin/curl
```

**Comment by satmandu** [comment]
> On armv7l, I get:
> 
> ```
> $ crew search curl -v
> curl: Command line tool and library for transferring data with URLs.
> https://curl.se/
> Version: 7.78.0
> License: curl
> $ curl -L# git.io/vddgY -o install.sh
> bash: /usr/local/bin/curl: No such file or directory
> $ ls -l /usr/local/bin/curl
> -rwxr-xr-x 1 chronos chronos 3514344 Aug 10 01:53 /usr/local/bin/curl
> ```

Thanks. debugging...

**Comment by uberhacker** [comment]
Also, on x86_64, I get:
```
$ curl -V
$ curl -L# git.io/vddgY -o install.sh
$ ls install.sh
ls: cannot access 'install.sh': No such file or directory
```
Doesn't say file not found but doesn't execute either, apparently.

**Comment by satmandu** [comment]
Looks like there's a musl dependency, which defeats the point of using musl, as we were using musl to get a static build.

(install musl to fix this, but this package build needs to be fixed @saltedcoffii )

**Comment by satmandu** [comment]
My initial thought would be that gold linker might break this...

File: packages/curl.rb (hunk 1 above)
Comment (saltedcoffii) [reply packages/curl.rb:74]: Honestly I've been thinking of switching entirely to musl with Chromebrew. Like wouldn't that fix so many glibc problems and allow chromebrew, and allow binaries to be much smaller and run faster? @satmandu let me know you're opinion, but I'd really like to start doing that.

**Comment by satmandu** [comment]
I'm redoing this based on the git package file, since that did produce a working static musl curl.

**Comment by satmandu** [comment]
FYI if testing this breaks curl, just do this:
`cp /usr/local/bin/curl_musl_static /usr/local/bin/curl`

**Comment by satmandu** [comment]
ok this is due to a bug in the 7.78.0 sources....

**Comment by satmandu** [comment]
Got it working on x86_64:
```
chronos@cheekon-x86_64 ~ $ curl --version
curl 7.78.0 (x86_64-pc-linux-gnu) libcurl/7.78.0 OpenSSL/1.1.1k zlib/1.2.11 brotli/1.0.9 zstd/1.5.0 c-ares/1.17.2 libidn2/2.3.2 nghttp2/1.44.0
Release-Date: 2021-07-21
Protocols: dict file ftp ftps gopher gophers http https mqtt pop3 pop3s smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets zstd
chronos@cheekon-x86_64 ~ $ ldd `which curl`
        not a dynamic executable
```

**Comment by satmandu** [comment]
Static builds should now be fixed. I imported from the git build, and then realized that this was a bug in 7.78.0, but sunk costs and all it now all appears to work in the containers.

Please see if this works on real hardware, please?

File: packages/curl.rb (hunk 1 above)
Comment (uberhacker) [reply packages/curl.rb:113]: This PR is way too complex.  It also won't build on i686.  Would it make sense to split it out into dependency packages?

Comment (satmandu) [reply packages/curl.rb:113]: Sure... That makes sense.

Let me create some musl_ packages, which will put stuff in the non-FHS compliant (but standard for musl!) #{CREW_PREFIX}/musl 

What do you think @saltedcoffii ?

**Comment by saltedcoffii** [comment]
Works on real x86_64 hardware as far as I can tell

File: packages/curl.rb (hunk 1 above)
Comment (saltedcoffii) [reply packages/curl.rb:113]: I think we have much bigger priorities. Many packages don't compile with musl without void or alpine patches, so easy patch downloading is a must, hence the extra url PR. I also think we need (and have needed for a while) a way to track and remove orphaned dependencies, much like `apt autoremove`. Is that reasonable @satmandu ?

**Comment by uberhacker** [comment]
It works fine on x86_64 for me too.

**Comment by uberhacker** [comment]
This also works on armv7l.  At least on my outdated release 84 EOL'd system.

**Comment by uberhacker** [comment]
@satmandu: Does this PR have any hope or should we start over?

**Comment by satmandu** [comment]
> @satmandu: Does this PR have any hope or should we start over?

I should have the curl 1.79.1 up shortly...

**Comment by satmandu** [comment]
7.79.1 is up here: https://github.com/skycocker/chromebrew/pull/6065

**Comment by uberhacker** [comment]
Superceded by #6065.

**Pull request closed without merging**